### PR TITLE
network: update ip in lb/pf/dnat tables when update vm nic ip

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes the issue that new ip is not working if old ip is used in lb/pf (or dnat).

Steps to reproduce the issue
(1) create isolated network, and vm
(2) acquire 3 public ips, add firewall rules to allow port 22 (ssh)
(3) create NAT to the vm on public ip 1, load balancer to vm port 22 (ssh) on public ip 2,  port forwarding rule to vm port 22 on public ip 3
(4) verify the ssh connection to vm, all work
(5) stop vm, and update vm nic ip, start vm
(6) expected result: DNAT, LB, PF work, 
actual result without this PR : LB and PF do not work.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -1748,35 +1749,33 @@ public UserVm updateNicIpForVirtualMachine(UpdateVmNicIpCmd cmd) {
...
+                    }
+                }
+            }
+            if (_networkModel.areServicesSupportedInNetwork(network.getId(), Service.PortForwarding)) {
+                List<PortForwardingRuleVO> firewallRules = _portForwardingDao.listByVm(vm.getId());
+                for (PortForwardingRuleVO firewallRule : firewallRules) {
+                    FirewallRuleVO rule = _rulesDao.findById(firewallRule.getId());
+                    if (nicVO.getIPv4Address().equals(firewallRule.getDestinationIpAddress().toString()) && network.getId() == rule.getNetworkId()) {
+                        firewallRule.setDestinationIpAddress(new Ip(ipaddr));
+                        _portForwardingDao.persist(firewallRule);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1777]: can we factor these out in twoo separate methods, please @weizhouapache ?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: Updated. thanks @DaanHoogland .

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm, tnx @weizhouapache

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -1747,38 +1748,12 @@ public UserVm updateNicIpForVirtualMachine(UpdateVmNicIpCmd cmd) {
...
-            // implementing the network elements and resources as a part of vm nic ip update if network has services and it is in Implemented state
-            if (!_networkModel.listNetworkOfferingServices(offering.getId()).isEmpty() && network.getState() == Network.State.Implemented) {
-                User callerUser = _accountMgr.getActiveUser(CallContext.current().getCallingUserId());
-                ReservationContext context = new ReservationContextImpl(null, null, callerUser, caller);
-                DeployDestination dest = new DeployDestination(_dcDao.findById(network.getDataCenterId()), null, null, null);
 
-                s_logger.debug("Implementing the network " + network + " elements and resources as a part of vm nic ip update");
-                try {
-                    // implement the network elements and rules again
-                    _networkMgr.implementNetworkElementsAndResources(dest, context, network, offering);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1766]: @weizhouapache is this call not required while updating nic?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1766]: @sureshanaparti 
it is not needed. 
the network rules in VR will be updated when vm is started.

Comment (sureshanaparti) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1766]: ok, thanks for the confirmation @weizhouapache

**Review by sureshanaparti — APPROVED** [review]

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -1826,6 +1801,45 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
...
+    private void updatePublicIpDnatVmIp(long vmId, long networkId, String oldIp, String newIp) {
+        if (_networkModel.areServicesSupportedInNetwork(networkId, Service.StaticNat)) {
+            List<IPAddressVO> publicIps = _ipAddressDao.listByAssociatedVmId(vmId);
+            for (IPAddressVO publicIp : publicIps) {
+                if (oldIp.equals(publicIp.getVmIp()) && publicIp.getAssociatedWithNetworkId() == networkId) {
+                    publicIp.setVmIp(newIp);
+                    _ipAddressDao.persist(publicIp);
+                }
+            }
+        }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1813]: We could invert the `if` logic and add a return to reduce indentation.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -1826,6 +1801,45 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
...
+            List<LoadBalancerVMMapVO> loadBalancerVMMaps = _loadBalancerVMMapDao.listByInstanceId(vmId);
+            for (LoadBalancerVMMapVO map : loadBalancerVMMaps) {
+                long lbId = map.getLoadBalancerId();
+                FirewallRuleVO rule = _rulesDao.findById(lbId);
+                if (oldIp.equals(map.getInstanceIp()) && networkId == rule.getNetworkId()) {
+                    map.setInstanceIp(newIp);
+                    _loadBalancerVMMapDao.persist(map);
+                }
+            }
+        }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1827]: We could invert the `if` logic and add a return to reduce indentation.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 5]
```diff
@@ -1826,6 +1801,45 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
...
+        if (_networkModel.areServicesSupportedInNetwork(networkId, Service.PortForwarding)) {
+            List<PortForwardingRuleVO> firewallRules = _portForwardingDao.listByVm(vmId);
+            for (PortForwardingRuleVO firewallRule : firewallRules) {
+                FirewallRuleVO rule = _rulesDao.findById(firewallRule.getId());
+                if (oldIp.equals(firewallRule.getDestinationIpAddress().toString()) && networkId == rule.getNetworkId()) {
+                    firewallRule.setDestinationIpAddress(new Ip(newIp));
+                    _portForwardingDao.persist(firewallRule);
+                }
+            }
+        }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1840]: We could invert the `if` logic and add a return to reduce indentation.

**Review by GutoVeronezi — COMMENTED** [review]
@weizhouapache I think it would be nice if we add some unit tests to these new methods.

**Comment by weizhouapache** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 3 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1813]: @GutoVeronezi
thanks for your suggestion. I will consider it if there are many code lines in the `if` block.
for these 3 methods, they have only 2 lines in `if` block, and I think `&&` is easier to understand than `!(aa) || !(bb)`

Comment (GutoVeronezi) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1813]: @weizhouapache sorry, selecting multiple lines may have confused my comment.

 I agree with you when we talk about the second `if` statement, however, I meant invert the first `if` statement, like this:

```suggestion
        if (!_networkModel.areServicesSupportedInNetwork(networkId, Service.StaticNat)) {
            return;
        } 

        List<IPAddressVO> publicIps = _ipAddressDao.listByAssociatedVmId(vmId);
        for (IPAddressVO publicIp : publicIps) {
            if (oldIp.equals(publicIp.getVmIp()) && publicIp.getAssociatedWithNetworkId() == networkId) {
                publicIp.setVmIp(newIp);
                _ipAddressDao.persist(publicIp);
            }
        }
```

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1813]: > @weizhouapache sorry, selecting multiple lines may have confused my comment.
> 
> I agree with you when we talk about the second `if` statement, however, I meant invert the first `if` statement, like this:
> 
> This way we can avoid the `hadouken code` =)

@GutoVeronezi ah ok, my misunderstanding. I will change it. thanks

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

I'm just missing unit tests.

**Pull request merged**